### PR TITLE
CheckEnv - Give new installs a grace period before 'Cron Not Running' msg

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Don't bug site admins with "Cron Not Running" error messages right after install. Instead, give them a day's grace period in which to set it up.

Inspired by discussion at https://lab.civicrm.org/dev/core/-/issues/1863#note_39918

Before
----------------------------------------
"Cron Not Running" error pops up immediately after installing CiviCRM.

After
----------------------------------------
The error is suppressed for 24 hrs.

Technical Details
----------------------------------------
The last cron run is recorded as a status preference. That's a little odd, but I'll go with it. 
Now if a cron run isn't recorded, we create one with `prefs = 'new'` to distinguish it from normal cron runs.
I'm honestly not sure what the 'prefs' column is *supposed* to be used for; grepping turns up nothing. But it seems to work for this purpose.

### CODE REVIEW

**Comment by colemanw** [comment]
@kcristiano I thought you'd appreciate this.

Also, while I was working on this I noticed some oddities in the StatusPreference BAO and cleaned it up: #17801

File: CRM/Utils/Check/Component/Env.php [hunk 1]
```diff
@@ -268,39 +269,50 @@ public function checkLastCron() {
...
+      $msg = ts('Last cron run at %1.', [1 => CRM_Utils_Date::customFormat(date('c', $statusPreference->check_info))]);
     }
+    // If cron record doesn't exist, this is a new install. Make a placeholder record (is_active=0).
     else {
-      $lastCron = 0;
-      $msg = ts('No cron runs have been recorded.');
+      $statusPreference = CRM_Core_BAO_StatusPreference::create([
+        'name' => 'checkLastCron',
+        'check_info' => $now,
+        'is_active' => 0,
```
Comment (seamuslee001) [inline CRM/Utils/Check/Component/Env.php:284]: should this not be set to `is_active = 1` and instead use the date expiry part of the check otherwise for new installs how will this be renabled?

File: CRM/Core/JobManager.php [hunk 2]
```diff
@@ -73,6 +73,7 @@ public function execute($auth = TRUE) {
     $statusPref = [
       'name' => 'checkLastCron',
       'check_info' => gmdate('U'),
+      'is_active' => 1,
```
Comment (colemanw) [inline CRM/Core/JobManager.php:76]: @seamuslee001 it gets reenabled here.

File: CRM/Utils/Check/Component/Env.php (hunk 1 above)
Comment (colemanw) [reply CRM/Utils/Check/Component/Env.php:284]: Oh, yes I see you're right it's a problem. I'll pick a different field.

**Comment by kcristiano** [comment]
@colemanw Thank You.

Great improvement.

`r-run`  worked as expected:

pop up:  
image

Status screen:  
image


No red, just lists of what to do.

File: CRM/Utils/Check/Component/Env.php [hunk 3]
```diff
@@ -258,49 +258,63 @@ public function checkDefaultMailbox() {
...
    * @return array
+   * @throws CRM_Core_Exception
    */
   public function checkLastCron() {
     $messages = [];
 
     $statusPreference = new CRM_Core_DAO_StatusPreference();
     $statusPreference->domain_id = CRM_Core_Config::domainID();
-    $statusPreference->name = 'checkLastCron';
+    $statusPreference->name = __FUNCTION__;
```
Comment (demeritcowboy) [inline CRM/Utils/Check/Component/Env.php:270]: Just curious why this change? The JobManager still refers to it as 'checkLastCron', so while it's unlikely somebody will change this function name, that would then get out of sync. Should it be a class constant which could then be used in both places?

Comment (colemanw) [reply CRM/Utils/Check/Component/Env.php:270]: @demeritcowboy the are actually the same thing, as the function name is "checkLastCron".

I had previously gotten confused about the relationship between the statusPreference and the function, and so I made this change to un-confuse myself. It's more consistent this way because the statusMessage generated below uses `__FUNCTION__` and not the string literal.

**Comment by colemanw** [comment]
@kcristiano thanks for the review. I've just made one additional change which is to switch from INFO to NOTICE when cron has not been set up. That seems more fitting, and will still not generate a popup warning.

**Comment by agh1** [comment]
I like the direction of this.  However, I am a little concerned that this creates a statusPreference, since if someone wanted to pre-hush the cron check (i.e. they know they'll never have cron running and they want to be sure the message doesn't pop up) they'd insert a statusPreference right after install.  I think this would undo their work.

**Comment by colemanw** [comment]
@agh1 this confused me at first too, but creating a statusPreference is actually not new to this PR.

The way the system currently works, a statusPreference with the name `"checkLastCron"` is created every time cron runs (and if one already exists, it gets updated) and it stores the current timestamp.

So all this PR does in addition to that is create (or update) a statusPreference with the placeholder timestamp `"new"` if a timestamp isn't found. The scenario you described where the install script creates one would work fine. This status check would update it with the placeholder timestamp and everything would work normally from there.

**Comment by agh1** [comment]
One thought is to just make the message a "notice" if no cron has ever run and no scheduled jobs (besides `version_check`) are enabled.  We can tell that they won't be missing out on much, so the severity is no longer "error: your mail is stuck and membership are eternal" but rather "notice: you haven't enabled something most people want".

**Comment by agh1** [comment]
@colemanw gotcha--I haven't looked at this in years and I think that used to be stored in a setting.

**Comment by agh1** [comment]
Rereading this my only thought is that the message during that grace period is a "notice", which is appropriate but means that nobody will see it unless they open up the status page.  We set up this check in the first place to replace a more ad hoc popup that appeared upon install.  My concern is that we will really need to emphasize in some other way that an admin should configure cron.

Otherwise, we're just deferring the shock and confusion of seeing "cron not running" until a day later.

**Comment by colemanw** [comment]
@agh1 ok I've downgraded the message from ERROR to WARNING in cases where cron has not ever run.

**Comment by agh1** [comment]
@colemanw yeah I think that's fair: it's more severe if you thought things were running than if you never thought it would.

Re: my prior comment about increasing visibility, I think a natural thing to do would be to have a message on the scheduled jobs page if cron isn't running, similar to what's there in a non-production environment:

https://github.com/civicrm/civicrm-core/blob/64aa560db1b9afecf26162220cf0eb8153ff5a11/CRM/Admin/Page/Job.php#L142-L144

I'd write it now, but I'd end up duplicating a lot of the code here in terms of logic.  Do you think it would make sense to abstract out some of this code so it's available as a sanity check elsewhere in CiviCRM?

**Comment by kcristiano** [comment]
@agh1 Regarding 
>  However, I am a little concerned that this creates a statusPreference, since if someone wanted to pre-hush the cron check (i.e. they know they'll never have cron running and they want to be sure the message doesn't pop up) they'd insert a statusPreference right after install. I think this would undo their work.

I tested using this snippet:

`wp civicrm api StatusPreference.create ignore_severity=critical name=checkLastCron`  

This hides the cron notice after the patch is applied. 

I agree adding a notice to the scheduled jobs page would be a big improvement.

**Comment by eileenmcnaughton** [comment]
@agh1 are your concerns address (ie do you consider this mergeable now)

**Comment by agh1** [comment]
@eileenmcnaughton Yeah I think this is good to merge.  In reading the code again I think there's not too much that would be duplicated in producing a warning for the scheduled jobs page, so you can ignore that last question in my previous comment.

**Comment by eileenmcnaughton** [comment]
OK - I read @kcristiano as being in favour too - thanks

**Comment by agh1** [comment]
See #17819 which adds a message at the top of the Scheduled Jobs page.
